src/signTxOutput.c

```diff
@@ -0,0 +1,664 @@
+#include "signTxOutput.h"
+#include "state.h"
+// #include "cardano.h"
+// #include "addressUtilsShelley.h"
+#include "uiHelpers.h"
+#include "signTxUtils.h"
+#include "uiScreens.h"
+#include "txHashBuilder.h"
+#include "textUtils.h"
+// #include "hexUtils.h"
+// #include "messageSigning.h"
+#include "bufView.h"
+#include "securityPolicy.h"
+
+// we want to distinguish the two state machines to avoid potential confusion:
+// ctx / subctx
+// stage / state
+// from ctx, we only make the necessary parts available to avoid mistaken overwrites
+static output_context_t* subctx = &(instructionState.signTxContext.stageContext.output_subctx);
+static common_tx_data_t* commonTxData = &(instructionState.signTxContext.commonTxData);
+static tx_hash_builder_t* txHashBuilder = &(instructionState.signTxContext.txHashBuilder);
+
+bool signTxOutput_isFinished()
+{
+	TRACE("Output submachine state: %d", subctx->state);
+	// we are also asserting that the state is valid
+	switch (subctx->state) {
+	case STATE_OUTPUT_FINISHED:
+		return true;
+
+	case STATE_OUTPUT_BASIC_DATA:
+	case STATE_OUTPUT_TOKEN_GROUP:
+	case STATE_OUTPUT_TOKEN_AMOUNT:
+	case STATE_OUTPUT_CONFIRM:
+		return false;
+
+	default:
+		ASSERT(false);
+	}
+}
+
+void signTxOutput_init()
+{
+	explicit_bzero(subctx, SIZEOF(*subctx));
+
+	subctx->state = STATE_OUTPUT_BASIC_DATA;
+}
+
+static inline void CHECK_STATE(sign_tx_output_state_t expected)
+{
+	TRACE("Output submachine state: current %d, expected %d", subctx->state, expected);
+	VALIDATE(subctx->state == expected, ERR_INVALID_STATE);
+}
+
+static inline void advanceState()
+{
+	TRACE("Advancing output state from: %d", subctx->state);
+
+	switch (subctx->state) {
+
+	case STATE_OUTPUT_BASIC_DATA:
+		if (subctx->numTokenGroups > 0) {
+			ASSERT(subctx->currentTokenGroup == 0);
+			subctx->state = STATE_OUTPUT_TOKEN_GROUP;
+		} else {
+			subctx->state = STATE_OUTPUT_CONFIRM;
+		}
+		break;
+
+	case STATE_OUTPUT_TOKEN_GROUP:
+		ASSERT(subctx->currentTokenGroup < subctx->numTokenGroups);
+
+		// we are going to receive token amounts for this group
+		ASSERT(subctx->numTokenAmounts > 0);
+
+		subctx->state = STATE_OUTPUT_TOKEN_AMOUNT;
+		break;
+
+	case STATE_OUTPUT_TOKEN_AMOUNT:
+		// we are done with the current token group
+		ASSERT(subctx->currentTokenAmount == subctx->numTokenAmounts);
```

Comment:
this sees a bit weird to me at first sight - I would expect the state machine to enter this state at the beginning of processing the sequence of token amounts for the respective token group - why isn't it that straightforward?